### PR TITLE
Workflows: Outlier Detection Blocks

### PR DESCRIPTION
**Pull request by yeldarby**
# Description

Adds two blocks that detect changes and outliers in a stream of embeddings.

## Type of change

-   [x] New feature (non-breaking change which adds functionality)

## How has this change been tested, please provide a testcase or example of how you tested the change?

Locally against my webcam & video feeds + added tests

## Any specific deployment considerations

No

## Docs

-   [x] Self-documenting via manifest

### CODE REVIEW

File: tests/workflows/unit_tests/core_steps/sampling/test_identify_outliers.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    assert result.get("warming_up")
+
+    # add a bit of variance
+    for i in range(32):
+        result = identify_changes_block.run(
+            **default_inputs,
+            embedding = get_perturbed_value(initial_value, 1e-6)
+        )
+
+    print(result)
```
Comment (hansent) [inline tests/workflows/unit_tests/core_steps/sampling/test_identify_outliers.py:41]:
```suggestion
```

File: inference/core/workflows/core_steps/sampling/identify_outliers/v1.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            embedding_normed = embedding / norm
+
+        self.samples += 1
+        warming_up = self.samples < warmup
+
+        # Store normalized embedding for vMF:
+        self.all_embeddings.append(embedding_normed)
+        if len(self.all_embeddings) > window_size:
+            # Remove oldest embedding
+            self.all_embeddings.pop(0)
```
Comment (hansent) [inline inference/core/workflows/core_steps/sampling/identify_outliers/v1.py:155]: probably no a big deal for window sizes we are dealing with, but I think if we're popping from beginning, `deque` might be better/faster data structure than a list.

File: inference/core/workflows/core_steps/sampling/identify_outliers/v1.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        # Determine outlier based on percentile thresholds
+        is_outlier = (percentile < threshold_percentile) or (
+            percentile > (1 - threshold_percentile)
+        )
+
+        # print(is_outlier, f"{round(percentile*100)}%", warming_up)
+
+        # We retain fields "average" and "std" for compatibility, though they are less meaningful
+        # now. We can still return the SMA/EMA-based averages/std for debugging, or just zeros.
+        return {
```
Comment (hansent) [inline inference/core/workflows/core_steps/sampling/identify_outliers/v1.py:193]: comment does match whats actually returned?

**Review by hansent — APPROVED** [review]
lgtm

I checked out the wikipedia pages and tried having chatGPT walk me through the math / von Mises distribution stuff.  Can't say I grok fully but seems right / makes sense based on the parts I can wrap my head around at least :P

File: inference/core/workflows/core_steps/sampling/identify_outliers/v1.py (hunk 3 above)
Comment (yeldarby) [reply inference/core/workflows/core_steps/sampling/identify_outliers/v1.py:193]: Oops no, I changed the output format when I removed those. Forgot the comment. Will update

**Comment by yeldarby** [comment]
> lgtm
> 
> I checked out the wikipedia pages and tried having chatGPT walk me through the math / von Mises distribution stuff. Can't say I grok fully but seems right / makes sense based on the parts I can wrap my head around at least :P

Me either tbh; @isaacrob-roboflow suggested it, o1 helped write it, and it seems to work pretty well.

**Review by hansent — APPROVED** [review]
re-approve after removing print as suggested

**Review by probicheaux — APPROVED** [review]

**Pull request merged**